File: src/com/eagle/programmar/Julia/Expressions/Julia_FunctionCall.java
// Copyright Eagle Legacy Modernization LLC, 2010-date
// Original author: Steven A. O'Hara, Jul 9, 2022

package com.eagle.programmar.Julia.Expressions;

import java.util.ArrayList;

import com.eagle.interpret.EagleInterpreter;
import com.eagle.interpret.EagleRunnable;
import com.eagle.interpret.EagleRunnableWithResult.Eagle_Statement_Result;
import com.eagle.math.EagleValue;
import com.eagle.programmar.Julia.Julia_Expression;
import com.eagle.programmar.Julia.Julia_Statement;
import com.eagle.programmar.Julia.Julia_Variable;
import com.eagle.programmar.Julia.Statements.Julia_Function;
import com.eagle.programmar.Julia.Symbols.Julia_Identifier_Reference;
import com.eagle.tokens.AbstractFunction;
import com.eagle.tokens.PrimaryOperator;
import com.eagle.tokens.SeparatedList;
import com.eagle.tokens.interfaces.AbstractExpression;
import com.eagle.tokens.interfaces.AbstractVariable;
import com.eagle.tokens.punctuation.PunctuationComma;
import com.eagle.tokens.punctuation.PunctuationLeftParen;
import com.eagle.tokens.punctuation.PunctuationRightParen;
import com.eagle.transform.EagleGenerator;
import com.eagle.transform.EagleTransformableExpression;
import com.eagle.transform.EagleTransformer;

public class Julia_FunctionCall extends PrimaryOperator
		implements EagleRunnable, EagleTransformableExpression
{
	public @S(10) Julia_Variable variable;
	public @S(20) PunctuationLeftParen leftParen;
	public @S(30) SeparatedList<Julia_Expression, PunctuationComma> argList;
	public @S(40) PunctuationRightParen rightParen;

	@Override
	public void interpret(EagleInterpreter interpreter)
	{
		Julia_Identifier_Reference id = variable.vars.first();
		String name = id.getValue();
		
		// See if it is builtin function
		if (name.equals("div") && argList.getPrimaryCount() == 2)
		{
			Julia_Expression numerExpr = argList.getPrimaryElement(0);
			Julia_Expression denomExpr = argList.getPrimaryElement(1);
			int numer = interpreter.getIntValue(numerExpr);
			int denom = interpreter.getIntValue(denomExpr);
			interpreter.pushInt(numer / denom);
			return;
		}
		if (name.equals("string"))
		{
			StringBuffer buff = new StringBuffer();
			for (int i = 0; i < argList.getPrimaryCount(); i++)
			{
				Julia_Expression expr = argList.getPrimaryElement(i);
				String val = interpreter.getStrValue(expr);
				buff.append(val);
			}
			interpreter.pushStr(buff.toString());
			return;
		}
		if (name.equals("startswith") && argList.getPrimaryCount() == 2)
		{
			Julia_Expression strExpr = argList.getPrimaryElement(0);
			Julia_Expression pattExpr = argList.getPrimaryElement(1);
			String str = interpreter.getStrValue(strExpr);
			String patt = interpreter.getStrValue(pattExpr);
			interpreter.pushBool(str.startsWith(patt));
			return;
		}
		if (name.equals("length") && argList.getPrimaryCount() == 1)
		{
			Julia_Expression strExpr = argList.getPrimaryElement(0);
			String str = interpreter.getStrValue(strExpr);
			interpreter.pushInt(str.length());
			return;
		}

		// Look up the function in our function list
		AbstractFunction fn = interpreter.findFunction(name);
		if (fn == null)
		{
			throw new RuntimeException("Unable to find a function named " + name);
		}
		Julia_Function func = (Julia_Function) fn;

		// Make sure the function args match up
		int argCount = argList.getPrimaryCount();
		int paramCount = func.params.parameters.getPrimaryCount();
		if (argCount != paramCount)
		{
			throw new RuntimeException(
					"Function " + name + " expects #args = " + paramCount + ", but was given " + argCount);
		}

		// Now assign all the parameters
		ArrayList<String> argTypes = new ArrayList<String>();
		for (int i = 0; i < argCount; i++)
		{
			Julia_Expression expr = argList.getPrimaryElement(i);
			Julia_Variable param = func.params.parameters.getPrimaryElement(i);

			EagleValue val = interpreter.getEagleValue(expr);
			interpreter.setSymbol(param, param.vars.first().getValue(), val);
			argTypes.add(val.typeName());
		}

		// Prepare to evaluate the method
		long startTime = System.nanoTime();

		// And transfer control to the method
		interpreter.callingFunction(name, func);
		Eagle_Statement_Result result = Eagle_Statement_Result.NORMAL;
		for (Julia_Statement stmt : func.stmts._elements)
		{
			result = interpreter.tryToInterpret(stmt);
			if (result != Eagle_Statement_Result.NORMAL) break;
		}

		// The result was already put on the runtime stack
		long elapsedTime = System.nanoTime() - startTime;
		func._callMetrics.addCallFrom(this, elapsedTime);
		func._argumentsMetrics.calledWith(argTypes);

		// Now remove all those parameters
		interpreter.completedFunction(name, func);
	}

	@Override
	public AbstractExpression transformExpression(EagleTransformer transformer,
			EagleGenerator generator)
	{
		Julia_Identifier_Reference id = variable.vars.first();
		String name = id.getValue();
		ArrayList<AbstractExpression> args = new ArrayList<AbstractExpression>();
		int argCount = argList.getPrimaryCount();
		for (int i = 0; i < argCount; i++)
		{
			Julia_Expression arg = argList.getPrimaryElement(i);
			AbstractExpression newArg = transformer.transformExpression(generator, arg);
			args.add(newArg);
		}

		AbstractVariable var = generator.newVariable(name);
		return generator.newMethodInvocation(var, args, id);
	}
}

File: csrc/com/eagle/programmar/Julia/Expressions/Julia_FunctionCall.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// Purchase a Premium Edition license at:
// https://www.tangiblesoftwaresolutions.com/order/order-java-to-csharp.html
// ====================================================================================================

using System;
using System.Collections.Generic;

// Copyright Eagle Legacy Modernization LLC, 2010-date
// Original author: Steven A. O'Hara, Jul 9, 2022

namespace com.eagle.programmar.Julia.Expressions
{

	using EagleInterpreter = com.eagle.interpret.EagleInterpreter;
	using EagleRunnable = com.eagle.interpret.EagleRunnable;
	using Eagle_Statement_Result = com.eagle.interpret.EagleRunnableWithResult.Eagle_Statement_Result;
	using EagleValue = com.eagle.math.EagleValue;
	using Julia_Expression = com.eagle.programmar.Julia.Julia_Expression;
	using Julia_Statement = com.eagle.programmar.Julia.Julia_Statement;
	using Julia_Variable = com.eagle.programmar.Julia.Julia_Variable;
	using Julia_Function = com.eagle.programmar.Julia.Statements.Julia_Function;
	using Julia_Identifier_Reference = com.eagle.programmar.Julia.Symbols.Julia_Identifier_Reference;
	using AbstractFunction = com.eagle.tokens.AbstractFunction;
	using PrimaryOperator = com.eagle.tokens.PrimaryOperator;
	using SeparatedList = com.eagle.tokens.SeparatedList;
	using AbstractExpression = com.eagle.tokens.interfaces.AbstractExpression;
	using AbstractStatement = com.eagle.tokens.interfaces.AbstractStatement;
	using AbstractType = com.eagle.tokens.interfaces.AbstractType;
	using AbstractVariable = com.eagle.tokens.interfaces.AbstractVariable;
	using PunctuationComma = com.eagle.tokens.punctuation.PunctuationComma;
	using PunctuationLeftParen = com.eagle.tokens.punctuation.PunctuationLeftParen;
	using PunctuationRightParen = com.eagle.tokens.punctuation.PunctuationRightParen;
	using EagleGenerator = com.eagle.transform.EagleGenerator;
	using TypeEnum = com.eagle.transform.EagleGenerator.TypeEnum;
	using EagleTransformableExpression = com.eagle.transform.EagleTransformableExpression;
	using EagleTransformer = com.eagle.transform.EagleTransformer;

	public class Julia_FunctionCall : PrimaryOperator, EagleRunnable, EagleTransformableExpression
	{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) com.eagle.programmar.Julia.Julia_Variable variable;
		public Julia_Variable variable;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) com.eagle.tokens.punctuation.PunctuationLeftParen leftParen;
		public PunctuationLeftParen leftParen;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(30) com.eagle.tokens.SeparatedList<com.eagle.programmar.Julia.Julia_Expression, com.eagle.tokens.punctuation.PunctuationComma> argList;
		public SeparatedList<Julia_Expression, PunctuationComma> argList;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(40) com.eagle.tokens.punctuation.PunctuationRightParen rightParen;
		public PunctuationRightParen rightParen;

		public override void interpret(EagleInterpreter interpreter)
		{
			Julia_Identifier_Reference id = variable.vars.first();
			string name = id.getValue();

			// Look up the function in our function list
			AbstractFunction fn = interpreter.findFunction(name);
			if (fn == null)
			{
				throw new Exception("Unable to find a function named " + name);
			}
			Julia_Function func = (Julia_Function) fn;

			// Make sure the function args match up
			int argCount = argList.getPrimaryCount();
			int paramCount = func.@params.parameters.getPrimaryCount();
			if (argCount != paramCount)
			{
				throw new Exception("Function " + name + " expects #args = " + paramCount + ", but was given " + argCount);
			}

			interpreter.callingFunction(name, func);

			// Now assign all the parameters
			List<EagleGenerator.TypeEnum> argTypes = new List<EagleGenerator.TypeEnum>();
			for (int i = 0; i < argCount; i++)
			{
				Julia_Expression expr = argList.getPrimaryElement(i);
				Julia_Variable param = func.@params.parameters.getPrimaryElement(i);

				EagleValue val = interpreter.getEagleValue(expr);
				interpreter.setSymbol(param, param.vars.first().getValue(), val);
				argTypes.Add(val.getType());
			}

			// Prepare to evaluate the method
			long startTime = System.nanoTime();

			// And transfer control to the method
			Eagle_Statement_Result result = Eagle_Statement_Result.NORMAL;
			foreach (Julia_Statement stmt in func.stmts._elements)
			{
				result = interpreter.tryToInterpret(stmt);
				if (result != Eagle_Statement_Result.NORMAL)
				{
					break;
				}
			}

			// The result was already put on the runtime stack
			long elapsedTime = System.nanoTime() - startTime;
			func._callMetrics.addCallFrom(this, elapsedTime);
			func._argumentsMetrics.calledWith(argTypes);

			// Now remove all those parameters
			interpreter.completedFunction(name, func);
		}

		public override AbstractExpression transformExpression(EagleTransformer transformer, EagleGenerator<AbstractStatement, AbstractExpression, AbstractVariable, AbstractType> generator)
		{
			Julia_Identifier_Reference id = variable.vars.first();
			string name = id.getValue();
			List<AbstractExpression> args = new List<AbstractExpression>();
			int argCount = argList.getPrimaryCount();
			for (int i = 0; i < argCount; i++)
			{
				Julia_Expression arg = argList.getPrimaryElement(i);
				AbstractExpression newArg = transformer.transformExpression(generator, arg);
				args.Add(newArg);
			}

			AbstractVariable var = generator.newVariable(name);
			return generator.newMethodInvocation(var, args, id);
		}
	}

}
